File: src/main/kotlin/school/charset/app/domain/exercise/generator/Utf16ExerciseGenerator.kt
package school.charset.app.domain.exercise.generator

import school.charset.app.domain.encoding.Codec
import school.charset.app.domain.encoding.Encoding
import school.charset.app.domain.exercise.Exercise
import school.charset.app.domain.exercise.ExerciseGenerationException
import school.charset.app.domain.exercise.Step

class Utf16ExerciseGenerator(
    private val codec: Codec,
    override val encoding: Encoding,
    private val codePointGenerator: CodePointGenerator,
    private val byteArrayGenerator: ByteArrayGenerator,
    private val utf16Generator: Utf16Generator,
) : EncodingExerciseGenerator {

    init {
        require(encoding == Encoding.Utf16Be || encoding == Encoding.Utf16Le) {
            "Utf16ExerciseGenerator requires Utf16Be or Utf16Le, got ${encoding.id}"
        }
    }

    private val endian: Encoding.Endian =
        if (encoding == Encoding.Utf16Be) Encoding.Endian.BigEndian else Encoding.Endian.LittleEndian

    override fun generateEncode(level: Int): Exercise.Encode {
        val utf16Level = parseLevel(level)
        val codePoint = codePointGenerator.randomUtf16(utf16Level)
        val steps = utf16Generator.buildEncodeStepsFor(codePoint, endian)
            .forExercise()
            .withOffsetStep(codePoint.value)
        return Exercise.Encode(codePoint, encoding, level, steps)
    }

    override fun generateDecode(level: Int): Exercise.Decode {
        val utf16Level = parseLevel(level)
        val bytes = byteArrayGenerator.randomUtf16(utf16Level, encoding)
        val codePoint = codec.decode(bytes, encoding)
        val steps = utf16Generator.buildDecodeStepsFor(bytes, codePoint, endian)
            .forExercise()
            .withDecodeOffsetStep(codePoint.value)
        return Exercise.Decode(bytes, codePoint, encoding, level, steps)
    }

    private fun List<Step>.forExercise(): List<Step> = withoutEndianness().simplifyBmp()

    // The exercise gives the target endianness in its header (a random BE/LE),
    // it isn't something the learner derives - so it's not a step. Byte order is
    // still tested via the hex-bytes step. The sandbox keeps the endianness step
    // (shared Utf16Generator) for its explanation panel.
    private fun List<Step>.withoutEndianness(): List<Step> = filterNot { it is Step.Endianness }

    // A BMP code point is a direct copy: the 16-bit code unit IS the scalar
    // value, so the binary step is just a hex<->binary detour. We drop it and go
    // straight from the format choice to the hex bytes. The surrogate case
    // (supplementary, identified by its bit-groups step) keeps binary + bit
    // groups, where the real packing happens.
    private fun List<Step>.simplifyBmp(): List<Step> = if (none { it is Step.BitGroups }) filterNot { it is Step.Binary } else this

    // Supplementary encode gets an explicit "subtract 0x10000" step (entered in
    // hex) right after the format choice, so the learner computes the 20-bit
    // value once and reads it back, instead of folding the subtraction into the
    // binary step. BMP (< 0x10000) has no offset, so it's left untouched.
    private fun List<Step>.withOffsetStep(codePointValue: Int): List<Step> {
        if (codePointValue < SUPPLEMENTARY_OFFSET) return this
        val afterFormat = indexOfFirst { it is Step.Format } + 1
        return toMutableList().apply {
            add(afterFormat, Step.Offset(expected = codePointValue - SUPPLEMENTARY_OFFSET))
        }
    }

    // Decode mirror: once the 20-bit scalar is assembled (binary step), an offset
    // step holds that value in hex, and the code-point step adds 0x10000. So the
    // "+ 0x10000" is explicit instead of folded into the code-point step.
    // Supplementary only; BMP has no binary step to follow.
    private fun List<Step>.withDecodeOffsetStep(codePointValue: Int): List<Step> {
        if (codePointValue < SUPPLEMENTARY_OFFSET) return this
        val afterBinary = indexOfFirst { it is Step.Binary } + 1
        return toMutableList().apply {
            add(afterBinary, Step.Offset(expected = codePointValue - SUPPLEMENTARY_OFFSET))
        }
    }

    private fun parseLevel(level: Int): Utf16Level = Utf16Level.fromNumber(level)
        ?: throw ExerciseGenerationException(
            encoding = encoding,
            level = level,
            reason = "level must be one of: ${Utf16Level.validNumbers}",
        )

    private companion object {
        private const val SUPPLEMENTARY_OFFSET = 0x10000
    }
}

File: lib/charset/exercise/generator/utf16_exercise_generator.ex
defmodule Charset.Exercise.Generator.Utf16ExerciseGenerator do
  @moduledoc """
  UTF-16 exercises (BE or LE), adapting the shared `Utf16Steps` builders:

  * the endianness step is dropped - the exercise gives the target byte order
    in its header (a random BE/LE), it isn't something the learner derives.
    Byte order is still tested via the hex-bytes step. The sandbox keeps the
    endianness step for its explanation panel.
  * BMP exercises drop the binary step - a BMP code point is a direct copy
    (the 16-bit code unit IS the scalar value), so the binary step is just a
    hex<->binary detour. The surrogate case keeps binary + bit groups, where
    the real packing happens.
  * supplementary exercises get an explicit "subtract/add 0x10000" offset
    step (entered in hex), so the conversion is its own step instead of being
    folded into the binary or code-point step.
  """

  alias Charset.Encoding
  alias Charset.Encoding.Codec
  alias Charset.Exercise
  alias Charset.Exercise.Generator.CodePointGenerator
  alias Charset.Exercise.Generator.Levels
  alias Charset.Exercise.Generator.Utf16Steps
  alias Charset.Exercise.GenerationError
  alias Charset.Exercise.Step

  @spec generate_encode!(:utf16be | :utf16le, pos_integer()) :: Exercise.t()
  def generate_encode!(encoding, level) when encoding in [:utf16be, :utf16le] do
    tier = parse_level!(encoding, level)
    code_point = CodePointGenerator.random_utf16(tier)

    Exercise.encode(code_point, encoding, level, exercise_encode_steps(code_point, encoding))
  end

  @spec generate_decode!(:utf16be | :utf16le, pos_integer()) :: Exercise.t()
  def generate_decode!(encoding, level) when encoding in [:utf16be, :utf16le] do
    tier = parse_level!(encoding, level)
    bytes = Codec.encode!(CodePointGenerator.random_utf16(tier), encoding)
    code_point = Codec.decode!(bytes, encoding)

    Exercise.decode(
      bytes,
      code_point,
      encoding,
      level,
      exercise_decode_steps(bytes, code_point, encoding)
    )
  end

  @doc "Deterministic exercise-flow step builder (exposed for tests)."
  @spec exercise_encode_steps(integer(), :utf16be | :utf16le) :: [Step.t()]
  def exercise_encode_steps(code_point, encoding) do
    code_point
    |> Utf16Steps.build_encode_steps_for(Encoding.endianness(encoding))
    |> for_exercise()
    |> with_encode_offset_step(code_point)
  end

  @doc "Deterministic exercise-flow step builder (exposed for tests)."
  @spec exercise_decode_steps(binary(), integer(), :utf16be | :utf16le) :: [Step.t()]
  def exercise_decode_steps(bytes, code_point, encoding) do
    bytes
    |> Utf16Steps.build_decode_steps_for(code_point, Encoding.endianness(encoding))
    |> for_exercise()
    |> with_decode_offset_step(code_point)
  end

  defp for_exercise(steps), do: steps |> without_endianness() |> simplify_bmp()

  defp without_endianness(steps) do
    Enum.reject(steps, &match?(%Step.Endianness{}, &1))
  end

  # A BMP exercise (no bit-groups step) goes straight from the format choice
  # to the hex bytes / code point.
  defp simplify_bmp(steps) do
    if Enum.any?(steps, &match?(%Step.BitGroups{}, &1)) do
      steps
    else
      Enum.reject(steps, &match?(%Step.Binary{}, &1))
    end
  end

  # Supplementary encode: the offset step (code point - 0x10000, in hex) goes
  # right after the format choice. BMP is left untouched.
  defp with_encode_offset_step(steps, code_point) do
    insert_offset_step(steps, code_point, Step.Format)
  end

  # Decode mirror: once the 20-bit scalar is assembled (binary step), the
  # offset step holds that value, and the code-point step adds 0x10000.
  defp with_decode_offset_step(steps, code_point) do
    insert_offset_step(steps, code_point, Step.Binary)
  end

  defp insert_offset_step(steps, code_point, after_struct) do
    offset = Utf16Steps.supplementary_offset()

    if code_point < offset do
      steps
    else
      index = Enum.find_index(steps, &is_struct(&1, after_struct)) + 1
      List.insert_at(steps, index, Step.Offset.new!(code_point - offset))
    end
  end

  defp parse_level!(encoding, level) do
    case Levels.parse(:utf16, level) do
      {:ok, tier} ->
        tier

      :error ->
        raise GenerationError,
          encoding: encoding,
          level: level,
          reason: "level must be one of: #{Levels.valid_numbers(:utf16)}"
    end
  end
end
